[Port] Adds sentient monkey helmets

XeonMations opened the pull request:
- [x] Fix runtimes during removal of helmets

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->
Adds in sentient monkey helmets, ordereable from cargo via the science tab.
Ported from: tgstation/tgstation/pull/51621

## Why It's Good For The Game

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/62395746/09d338fd-3704-46ce-a4ae-91f41f5b9c73)


## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>


https://github.com/BeeStation/BeeStation-Hornet/assets/62395746/a364673f-e4a9-4ee7-8baf-94e46e8fa932



</details>

## Changelog
:cl: XeonMations, Dakae
add: Sentient monkey helmets, orderable from cargo
refactor: TRAIT_MONKEYLIKE is now TRAIT_DISCOORDINATED
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

XeonMations committed:
Helmets

Yet to fix runtimes during removal due to qdel being referenced multiple times

Tsar-Salat commented on code/modules/clothing/head/helmet.dm:
Unique helmet like this has enough extra functionalities tacked on that it really shouldnt be in `code/modules/clothing/head/helmet.dm`, recommend putting it in its own file, like `code/modules/clothing/head/mind_monkey_helmet.dm`

XeonMations committed:
Monkey 2

Changed to seperate file, yet to fix the runtime when removing headset from own head as a monkey

XeonMations committed:
Fixed runtime issues

Fixed runtime issue when removing own helmet as a monkey, as well as adding a clause to delete the helmet itself if you somehow manage to remove a helmet during polling

itsmeow reviewed (approved):
Sprite is not great, but the code is fine.

Tsar-Salat reviewed (approved):
dunno why I was requested for review, but yeah it looks good

PowerfulBacon reviewed (dismissed):
At least put a small amount of effort into the sprite, or find someone to sprite it (Or copy an existing sprite and make it look more wire-y).

HowToLoLu committed:
Adds a timer to the removal of the helmet

HowToLoLu committed:
fixes the icon so it doesn't disappear

HowToLoLu committed:
Small grammar change

HowToLoLu committed:
Prefs menu stuff and proper poll ignore

XeonMations committed:
Merge pull request #1 from HowToLoLu/confirmation-timer

Confirmation timer

XeonMations committed:
Updated polling check code to make it look cleaner

HowToLoLu committed:
visible_messages, eh?

HowToLoLu committed:
Antispam is actually now antispam

HowToLoLu committed:
we're getting there...

HowToLoLu committed:
Should've refectored the type a while ago

HowToLoLu committed:
refactor that a bit...

HowToLoLu committed:
Well, forgot about the pack

HowToLoLu committed:
Spliting up disconnet

HowToLoLu committed:
Weakening the magnification reference

HowToLoLu committed:
Knock that off

HowToLoLu committed:
Whoopsie

HowToLoLu committed:
Even more split

HowToLoLu committed:
Maybe not do that?

HowToLoLu committed:
Makes a few comments more up to date

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,107 @@
+	if (magnification)
+		icon_state = "[base_icon_state]up"
+		return
+	icon_state = "[base_icon_state]"
```
```suggestion
	else
		icon_state = "[base_icon_state]"
```

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,107 @@
+		to_chat(something, "<span class='boldnotice'>You feel a stabbing pain in the back of your head for a moment.</span>")
+		something.apply_damage(5,BRUTE,BODY_ZONE_HEAD,FALSE,FALSE,FALSE) //notably: no damage resist (it's in your helmet), no damage spread (it's in your helmet)
+		playsound(src, 'sound/machines/buzz-sigh.ogg', 30, TRUE)
+		return
```
after this part, the codes should be in a separated proc, and be called by `INVOKE_ASYNC`
Doing this without async means `equipped()` is damn delayed but that shouldn't be a thing.

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,107 @@
+	strip_delay = 100
+	var/cooldown_expiry //It'll get annoying quick when someone tries to remove their own helmet 20 times a second
+	var/mob/living/carbon/monkey/magnification = null ///if the helmet is on a valid target (just works like a normal helmet if not (cargo please stop))
+	var/polling = FALSE///if the helmet is currently polling for targets (special code for removal)
```
this variable is not necessary if you write the logic well.

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,107 @@
+	playsound(src, 'sound/machines/ping.ogg', 30, TRUE)
+	polling = TRUE
+	var/list/candidates = pollCandidatesForMob("Do you want to play as a mind magnified monkey?", ROLE_MONKEY_HELMET, null, ROLE_MONKEY_HELMET, 50, magnification, POLL_IGNORE_MONKEY_HELMET)
+	polling = FALSE
```
remove polling variables, and put check here `if( the item is on this monkey) - if not, return`
This is why I said polling isn't necessary.

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,107 @@
+		magnification = null
+		visible_message("<span class='notice'>[src] falls silent. Maybe you should try again later?</span>")
+		playsound(src, 'sound/machines/buzz-sigh.ogg', 30, TRUE)
+	else
```
```suggestion
	else if(magnification) // Even once a poll is concluded, the monkey still should wear the helmet.
		if(magnification.key) // If multiple polls happen, but a ghost won from a first poll, they won't be replaced.
			return
```
This is why you don't need `polling`

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,107 @@
+/obj/item/clothing/head/mind_monkey_helmet/monkey_sentience/proc/disconnect()
+	if(!magnification) //not put on a viable head
+		return
+	if(polling)//put on a viable head, but taken off after polling finished.
```
this if check will not be necessary.

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,107 @@
+		new /obj/effect/decal/cleanable/ash/crematorium(drop_location()) //just in case they're in a locker or other containers it needs to use crematorium ash, see the path itself for an explanation
+		qdel(src)
+	else
+		if(magnification.client) //either used up correctly or taken off before polling finished (punish this by destroying the helmet)
```
do not use client. use key.
client can be null when the user is disconnected.

HowToLoLu committed:
Major Redux 2: Mind Swap Boogaloo

HowToLoLu committed:
realized a small flaw

HowToLoLu committed:
I messed up using signals this whole time...

HowToLoLu committed:
Two things one commit (icon change, and one misc)

HowToLoLu committed:
obj version now here, and mob version named right

HowToLoLu committed:
MAYBE I GOT IT THIS TIME

HowToLoLu committed:
Well that was pointless

HowToLoLu committed:
Hopefully this reads a bit clearer now

XeonMations committed:
Merge pull request #2 from HowToLoLu/monkey-helm-fixes-and-polishes

Monkey Helm 2: Mind Hat Boogaloo

XeonMations committed:
Reverts accidental void cloak change

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -32,76 +17,90 @@
-	var/list/candidates = pollCandidatesForMob("Do you want to play as a mind magnified monkey?", ROLE_MONKEY_HELMET, null, ROLE_MONKEY_HELMET, 50, magnification, POLL_IGNORE_MONKEY_HELMET)
-	polling = FALSE
+	var/list/candidates = pollCandidatesForMob("Do you want to play as a mind magnified monkey?", ROLE_MONKEY_HELMET, null, ROLE_MONKEY_HELMET, 50, user, POLL_IGNORE_MONKEY_HELMET)
+	//Some time has passed, and we could've been disintegrated for all we know (especially if we touch touch supermatter)
```
need `if(magnification) return` here so that it won't be retaken by someone when two polls exist.

EvilDragonfiend replied:
also, need another if check `if(src == user.head)`. You'll need another variable to get head variable.

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,106 @@
+	icon_state = "monkeymind"
+	strip_delay = 100
+	clothing_flags = EFFECT_HAT
+	var/cooldown_expiry //It'll get annoying quick when someone tries to remove their own helmet 20 times a second
```
use COOLDOWN defines

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,106 @@
+		return ..() //In case the monkey was already sentient
+
+	//Spam? No thanks, we're good.
+	if(cooldown_expiry <= world.time)
```
use COOLDOWN defines here too

HowToLoLu committed:
Cooldowns: Now with 100% more define usage

HowToLoLu committed:
Poll now has an additional check as suggested

XeonMations committed:
Merge pull request #3 from HowToLoLu/monkey-helm-fixes-and-polishes

Monkey helm fixes and polishes

EvilDragonfiend commented on code/__DEFINES/role_preferences.dm:
```diff
@@ -98,6 +99,7 @@ GLOBAL_LIST_INIT(special_roles, list(
 	ROLE_SPACE_PIRATE,
 	ROLE_FUGITIVE,
 	ROLE_FUGITIVE_HUNTER,
+	ROLE_MONKEY_HELMET,
```
Remove it from here.
We're planning to change midround popup to be shown to everyone always (until they set it poll-ignore)

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,106 @@
+		current = magnification.current //In case we weren't called by COMSIG_MIND_TRANSFER_TO
+	magnification = null
+	to_chat(current, "<span class='userdanger'>You feel your flicker of sentience ripped away from you, as everything becomes dim...</span>")
+	current.ghostize(FALSE)
```
put a comment `// Force-DNR`

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,106 @@
+/obj/item/clothing/head/monkey_sentience_helmet/proc/poll(mob/living/carbon/monkey/user) //At this point, we can assume we're given a monkey, since this'll put them in the body anyways
+	user.visible_message("<span class='warning'>[src] powers up!</span>")
+	playsound(src, 'sound/machines/ping.ogg', 30, TRUE)
+	var/list/candidates = pollCandidatesForMob("Do you want to play as a mind magnified monkey?", ROLE_MONKEY_HELMET, null, ROLE_MONKEY_HELMET, 50, user, POLL_IGNORE_MONKEY_HELMET)
```
```suggestion
	var/list/candidates = pollCandidatesForMob(
		Question = "Do you want to play as a mind magnified monkey?", 
		jobbanType = ROLE_MONKEY_HELMET, 
		gametypeCheck = null, 
		be_special_flag = null,
		poll_time = 100, 
		M = user, 
		ignore_category = POLL_IGNORE_MONKEY_HELMET)
```
to be honest, this proc is bad to read. Specified arguments will be better.
Also, `be_special_flag` can be null because we're going to make it always visible.
and 5 seconds are too short

EvilDragonfiend reviewed (approved):
Conditional approval: after my change request is addressed.
Code is fine.

EvilDragonfiend commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,106 @@
+		playsound(src, 'sound/machines/buzz-sigh.ogg', 30, TRUE)
+		return
+	var/mob/picked = pick(candidates)
+	user.key = picked.key
```
hmm, I am not sure if this is the correct way to make a new mind... please check if their mind is fine.

HowToLoLu replied:
According to a comment at the top of `code/datums/mind.dm`, this is already handled by `/mob/Login()`, which runs `/mob/sync_mind()`, and then runs `/mob/mind_initialize()`, where the key is set.

XeonMations committed:
Refactored polling + removed from role preferences

XeonMations committed:
Updated name + desc, removed examine proc

XeonMations committed:
Fixed singular comma in special_roles list

XeonMations committed:
Fixed runtime issue with ghostize

HowToLoLu committed:
Merge commit 'c26b440e834695851f34d5ea8a3220146ed7936a' into xeonmations-monkey-sentience-hat

HowToLoLu committed:
Merge commit '8c482abfe4f904c588713509327926c5d1d368aa' into xeonmations-monkey-sentience-hat

XeonMations committed:
Bring the PR up to date with current beecode

Avoids alot of conflicts

XeonMations committed:
Updated poll for mobs to work with meowdev changes, adds check to see if monkey is alive

HowToLoLu committed:
New sprite, who this?

Rukofamicom commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,108 @@
+	magnification = user.mind
+	RegisterSignal(magnification, COMSIG_MIND_TRANSFER_TO, PROC_REF(disconnect))
+	playsound(src, 'sound/machines/microwave/microwave-end.ogg', 100, FALSE)
+	to_chat(user, "<span class='notice'>You're a mind magnified monkey! Protect your helmet with your life- if you lose it, your sentience goes with it!</span>")
```
```suggestion
	to_chat(user, "<span class='notice'>You're a mind magnified monkey! Protect your helmet with your life; if you lose it, your sentience goes with it! Your helmet also strongly compels you to assist Nanotrasen and you should always act with the best interests of the station in mind.</span>")
	to_chat(user, "<span class='warning'></span>")
```

Rukofamicom reviewed (changes requested):
After some discussion with Admins, there are actually three changes I'd like to request to make sure this is a good fit for the server and it doesn't cause them unnecessary headaches:

1) Monkeys should have a clearly defined allegiance to the station, due to the helmets being provided through cargo by Nanotrasen.

2) The helmet should grant not just sentience but also dexterity - this will allow monkeys to use computers and other items that otherwise would require a more human touch. This will fully enable the monkeys to behave as humans and curb the desire to be humanized.

3) Attempting to transform a monkey with one of these helmets should remove or destroy the helmet before the transformation to prevent this from being a shortcut to adding fully fledged humans back into the round.

XeonMations committed:
Changed monkey allegiance text

XeonMations committed:
Merge branch 'monkey-sentience-hat' into new_msh_sprite

XeonMations committed:
Merge pull request #5 from HowToLoLu/new_msh_sprite

New Sprite + Description Update

XeonMations committed:
"It should be simple to replicate yourself"

XeonMations committed:
Like so?

XeonMations committed:
now?

HowToLoLu committed:
Refactors Discoordinated Trauma Trait

HowToLoLu committed:
Something about monkeys on typewriters...

HowToLoLu committed:
more tweaks to the item itself? sign me up!

XeonMations committed:
Fixed unexisting variable code.

XeonMations committed:
Merge pull request #6 from XeonMations/dakaechanges

"It should be simple to replicate yourself"

XeonMations committed:
Merge pull request #7 from HowToLoLu/hand-monkey-coordination

Hand-Monkey-Coordination

itsmeow commented on code/modules/clothing/head/mind_monkey_helmet.dm:
```diff
@@ -0,0 +1,122 @@
+		role_preference_key = null,
+		poll_time = 100,
+		M = user,
+		ignore_category = POLL_IGNORE_MONKEY_HELMET)
```
this pollignore is not necessary since the bantype is not used by anything else. poll ignores are autogenerated from the bantype or role preference.

itsmeow reviewed (approved):
The code looks good and the sprite is really just fine, I think it looks good. Allegiance now is good, being NT aligned. Maybe if you emag it or something you could get a syndicate one, but that could be later.

Rukofamicom reviewed (approved):
I rescind the request for total dexterity as part of the helmet. This can be potentially added at a later time and is not actually necessary for this role to be fun or useful. The partial "dexterity" added is good enough for now. 

Would still be neat if you (or anyone else) wants to give the monkeys access to consoles. *I was not aware of how complex the system was when I requested this from a first time contributor*
